src/compute-client/src/controller/instance.rs

```diff
@@ -1977,13 +2003,22 @@ where
 
             let compute_frontiers = collection.compute_dependencies.iter().flat_map(|dep_id| {
                 let collection = self.compute.collections.get(dep_id);
-                collection.map(|c| &c.write_frontier)
-            });
-            let storage_frontiers = collection.storage_dependencies.iter().flat_map(|dep_id| {
-                let collection = &self.storage_controller.collection(*dep_id).ok();
-                collection.map(|c| &c.write_frontier)
+                collection.map(|c| c.write_frontier.clone())
             });
 
+            let existing_storage_dependencies = collection
+                .storage_dependencies
+                .iter()
+                .filter(|id| self.storage_controller.check_exists(**id).is_ok())
+                .cloned()
```

Comment:
Nit: For `Copy` objects I'd prefer `copied()` instead of `cloned()`. Makes it obvious that we are not doing expensive allocations here and alerts us when we attempt to start doing so.

Reply:
TIL: I didn't know that `copied()` existed!